src/persist/src/s3.rs

```diff
@@ -468,36 +476,75 @@ impl Blob for S3Blob {
 
                 // Request the body.
                 let body_start = Instant::now();
-                let mut body = Vec::new();
+                let mut body_parts = Vec::new();
+
+                // Get the data into lgalloc at the absolute earliest possible
+                // point without (yet) having to fork the s3 client library.
+                let enable_s3_lgalloc = if self.is_cc_active {
+                    ENABLE_S3_LGALLOC_CC_SIZES.get(&self.cfg)
+                } else {
+                    ENABLE_S3_LGALLOC_NONCC_SIZES.get(&self.cfg)
+                };
+
+                // Ideally we write all of the copy all of the bytes into a
+                // single allocation, but we retain a CYA fallback case.
+                let enable_one_allocation =
+                    ENABLE_ONE_ALLOC_PER_REQUEST.get(&self.cfg) && enable_s3_lgalloc;
+                let mut buffer = match object.content_length() {
+                    Some(len @ 1..) if enable_one_allocation => {
+                        let len: u64 = len.try_into().expect("positive integer");
+                        let buf: MetricsRegion<u8> = self
+                            .metrics
+                            .lgbytes
+                            .persist_s3
+                            .new_region(usize::cast_from(len));
```

Comment:
IIUC this lets lgalloc decide if it's "too small" to go into lgalloc, whereas before we tried hard to get everything into lgalloc. Might be a good idea, but it's a decent behavior change between the two. I'd lean toward the previous behavior to derisk the PR a bit.

Reply:
Good call! I updated the `new_region(...)` method so we try allocating something that is at least as large as the smallest `lgalloc` class size. It should match exactly the behavior of the existing `try_mmap_region(...)` method now.